tool/lua/embed/init.lua

```diff
@@ -0,0 +1,451 @@
+  -- Load the rockspec in our environment
+  local chunk, err = load(content, "rockspec", "t", env)
+  if not chunk then
+    errorf("Failed to parse rockspec: %s", err)
+  end
+
+  local ok, result = pcall(chunk)
+  if not ok then
+    errorf("Failed to execute rockspec: %s", result)
+  end
+
+  return env
+end
+
+local function find_rock_download_url(rockspec_data, author, package_name, version)
+  -- Check source.url in rockspec
+  if rockspec_data.source and rockspec_data.source.url then
+    return rockspec_data.source.url
+  end
+
+  -- Try standard LuaRocks rock location
+  local rock_url = string.format("%s/manifests/%s/%s-%s.%s.rock",
+                                 LUAROCKS_API, author, package_name, version, "all")
+  return rock_url
+end
+
+--------------------------------------------------------------------------------
+-- Package extraction
+--------------------------------------------------------------------------------
+
+local function extract_lua_files_from_zip(zip_content, package_name)
+  -- Write to temp file
+  local tmpfile = "/tmp/cosmo-embed-" .. package_name .. ".zip"
+  local fd = unix.open(tmpfile, unix.O_WRONLY | unix.O_CREAT | unix.O_TRUNC, 0644)
+  if not fd then
+    errorf("Failed to create temp file: %s", tmpfile)
+  end
+  unix.write(fd, zip_content)
+  unix.close(fd)
+
+  -- Open ZIP
+  local reader, err = zip.open(tmpfile)
+  if not reader then
+    unix.unlink(tmpfile)
+    errorf("Failed to open ZIP: %s", err)
+  end
+
+  -- Extract all .lua files
+  local files = {}
+  local entries = reader:list()
+
+  for _, entry in ipairs(entries) do
+    if entry:match("%.lua$") and not entry:match("/$") then
+      local content, err = reader:read(entry)
+      if content then
+        files[entry] = content
+      else
+        log("Warning: Failed to read " .. entry .. ": " .. tostring(err))
+      end
+    end
+  end
+
+  reader:close()
+  unix.unlink(tmpfile)
+
+  return files
+end
+
+local function normalize_paths(files, package_name)
+  -- Map extracted paths to /zip/.lua/ structure
+  local normalized = {}
+
+  for path, content in pairs(files) do
+    local zippath = path
+
+    -- Remove common prefixes
+    zippath = zippath:gsub("^[^/]+/lua/", "")  -- package-1.0/lua/ -> ""
+    zippath = zippath:gsub("^lua/", "")         -- lua/ -> ""
+
+    -- If path doesn't start with package name, prepend it
+    if not zippath:match("^" .. package_name .. "/") and
+       not zippath:match("^" .. package_name .. "%.lua$") then
+      zippath = package_name .. "/" .. zippath
+    end
+
+    -- Ensure .lua/ prefix for ZIP
+    if not zippath:match("^%.lua/") then
+      zippath = ".lua/" .. zippath
+    end
+
+    normalized[zippath] = content
+  end
+
+  return normalized
+end
+
+--------------------------------------------------------------------------------
+-- Main embed functions
+--------------------------------------------------------------------------------
+
+function embed.install(package_name, output_path)
```

Comment:
can we copy and append files directly from the src to our destination without the temp files in the middle?